### PR TITLE
Enhance $expand clause to support nested filter

### LINKED ISSUE #637: simple odata client filter nested collection
I have the following odata query:

`/Places?$filter=CountryCode eq 'de'&$expand=Images($filter=IsCover eq true)`

The Places has a collection of Images. This query returns back all places for "DE" and for each place only the Image that has the property IsCover true

How can I write this in fluent API?

I tried following:

`return await _oDataClient .For<Place>().Filter(p => p.CountryCode == "DE")).Expand(f => f.Images.Where(i => i.IsCover == true)).FindEntriesAsync();`

but it keeps giving an exception (unsupported expression). Is there any way to write this in fluent API? I'm using the latest version of simple.odata.client (5.8.0)

### PR DESCRIPTION
**Pull request by maxbruecken**
This pull request adds support for nested filter while expanding collections. This should fix #637.

### CODE REVIEW

File: src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs [hunk 1]
```diff
@@ -331,7 +331,7 @@ public async Task ExpandCategorySelectProductNameCategoryName(string metadataFil
 
         [Theory]
         [InlineData("Northwind3.xml", "Products?$expand=*&$select=ProductName,Category/CategoryName")]
-        [InlineData("Northwind4.xml", "Products?$expand=*($select=CategoryName)&$select=ProductName")]
+        [InlineData("Northwind4.xml", "Products?$expand=*&$select=ProductName")]
```
Comment (watfordgnf) [inline src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs:334]: Is that expand+select syntax no longer supported? And if so, is that a breaking change we'll need to communicate?

File: src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs [hunk 2]
```diff
@@ -467,8 +467,8 @@ public async Task ExpandFunctionMultipleLevelsWithSelect()
                 "$expand=Product($expand=ProductCategory($expand=Category($expand=CategorySalesArea;$select=Code);$select=IsPrimary);$select=Id,ManufacturerId)," +
                 "ClientProductSkuPriceList($select=CurrencyId)," +
                 "ClientProductSkuSalesArea," +
-                "Product($expand=SupplierProductSkuClient($expand=SupplierProductSku($expand=SupplierProductSkuPriceList($expand=SupplierPriceList)));$select=Id,ManufacturerId)," +
-                "Product($expand=SupplierProductSkuClient($expand=SupplierProductSku($expand=SupplierProductSkuOnHand($expand=Warehouse)));$select=Id,ManufacturerId)&" +
+                "Product($expand=SupplierProductSkuClient($expand=SupplierProductSku($expand=SupplierProductSkuPriceList($expand=SupplierPriceList)));$expand=ProductCategory($expand=Category($select=Code);$select=IsPrimary);$select=Id,ManufacturerId)," +
+                "Product($expand=SupplierProductSkuClient($expand=SupplierProductSku($expand=SupplierProductSkuOnHand($expand=Warehouse)));$expand=ProductCategory($expand=Category($select=Code);$select=IsPrimary);$select=Id,ManufacturerId)&" +
```
Comment (watfordgnf) [inline src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs:471]: Why was this necessary? Is this a breaking change?

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs [hunk 3]
```diff
@@ -106,19 +98,52 @@ protected override void FormatInlineCount(IList<string> commandClauses)
...
+                            .Where(
+                                a => _session.Metadata.HasNavigationProperty(associatedEntityCollection.Name, a.Name))
+                            .Select(a =>
+                                FormatExpansionSegment(a, associatedEntityCollection, ODataExpandOptions.ByValue(),
+                                    command,
+                                    false))
+                            .Select(formattedExpand => $"{ODataLiteral.Expand}={formattedExpand}"));
+                    }
+
+                    var selectColumns = string.Join(",", association.ExpandAssociations.ToList()
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:132]: Is the embedded `ToList` required here? `String.Join` will execute this LINQ expression immediately and should not require coercion.

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs [hunk 4]
```diff
@@ -106,19 +98,52 @@ protected override void FormatInlineCount(IList<string> commandClauses)
...
+                                    !_session.Metadata.HasNavigationProperty(associatedEntityCollection.Name, a.Name))
+                        .Select(a => a.Name));
+                    if (!string.IsNullOrEmpty(selectColumns))
+                        clauses.Add($"{ODataLiteral.Select}={selectColumns}");
+                }
+
+                if (expandsToCollection && association.OrderByColumns.Any())
+                {
+                    var columns = string.Join(",", association.OrderByColumns
+                        .Select(o => o.Name + (o.Descending ? " desc" : string.Empty)).ToList());
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:143]: Is the `ToList` required here?

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs [hunk 5]
```diff
@@ -127,84 +152,68 @@ protected override void FormatInlineCount(IList<string> commandClauses)
...
-            if (string.IsNullOrEmpty(path))
+            if (string.IsNullOrEmpty(orderByColumn.Key))
+                return expandAssociation;
+            
+            var segments = orderByColumn.Key.Split('/');
+            if (segments[0] != expandAssociation.Name)
+                return expandAssociation;
+
+            var result = expandAssociation.Clone();
+            MergeOrderByColumns(result, segments.Skip(1).ToArray(), orderByColumn.Value);
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:165]: This should be skipped if segments.Length == 1, avoiding the LINQ projection and ToArray call.

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs [hunk 6]
```diff
@@ -127,84 +152,68 @@ protected override void FormatInlineCount(IList<string> commandClauses)
...
+            return result;
+        }
+
+        private static void MergeOrderByColumns(ODataExpandAssociation expandAssociation,
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:169]: This should be iterative if at all possible as opposed to recursive. On the hot path this would be a lot of Skip/ToArray calls.

File: src/Simple.OData.Client.Core/Expressions/ColumnExpression.cs [hunk 7]
```diff
@@ -36,7 +35,7 @@ public static string ExtractColumnName(this Expression expression, ITypeCache ty
             return expression.ExtractColumnNames(typeCache).Single();
         }
 
-        private static IEnumerable<string> ExtractColumnNames(this Expression expression, ITypeCache typeCache)
+        public static IEnumerable<string> ExtractColumnNames(this Expression expression, ITypeCache typeCache)
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Expressions/ColumnExpression.cs:38]: because the containing class ColumnExpression is internal, it shouldn't matter that this method is public.

[the remark this answers is not in the document]
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Expressions/ColumnExpression.cs:38]: I missed that, my apologies!

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 8]
```diff
@@ -333,52 +333,32 @@ public FluentCommand Expand(ODataExpandOptions expandOptions)
...
-            _details.ExpandAssociations.AddRange(SplitItems(associations).Select(x => new KeyValuePair<string, ODataExpandOptions>(x, ODataExpandOptions.ByValue())));
-            return this;
-        }
-
-        public FluentCommand Expand(ODataExpandOptions expandOptions, IEnumerable<string> associations)
-        {
-            if (IsBatchResponse) return this;
-
-            _details.ExpandAssociations.AddRange(SplitItems(associations).Select(x => new KeyValuePair<string, ODataExpandOptions>(x, expandOptions)));
+            _details.ExpandAssociations.AddRange(new[]
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs:336]: you're right, I will siplify this method

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs (hunk 4 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:143]: You're right, the .ToList() call is not necessary here. I will remove this call.

File: src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs (hunk 1 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs:334]: Expand+select is supported but only for concrete navigation properties e.g. $expand=Category($select=CategoryName). According to OData v4 specification (at least as implemented by Microsoft) a star operator (expand all navigation properties) can be combined only with $levels clause for cyclic properties. Because of star operator the server response will include Category/CategoryName.

File: src/Simple.OData.Client.Core/Adapter/CommandFormatterBase.cs [hunk 9]
```diff
@@ -298,5 +298,18 @@ protected string FormatOrderByItem(KeyValuePair<string, bool> pathWithOrder, Ent
                 commandClauses.Add($"{clauseLiteral}={string.Join(",", clauses.Select(x => formatItem(x, entityCollection)))}");
             }
         }
+
+        protected static IEnumerable<KeyValuePair<string, ODataExpandOptions>> FlatExpandAssociations(
+            IEnumerable<KeyValuePair<ODataExpandAssociation, ODataExpandOptions>> associations)
+        {
+            return associations
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Adapter/CommandFormatterBase.cs:305]: In my last commit I have replaced the recursion by expression visitor.

File: src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                    association.FilterExpression = association.FilterExpression && filterExpression;
+                return associations;
+            }
+
+            throw Utils.NotSupportedExpression(callExpression);
+        }
+
+        private static IEnumerable<ODataExpandAssociation> ExtractExpandAssociations(NewExpression newExpression, ITypeCache typeCache)
+        {
+            return newExpression.Arguments.SelectMany(x => ExtractExpandAssociations(x, typeCache));
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs]: After replacing the recursion by expression visitor the expand association for each argument will be added immediatelly to the result associations.

File: src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        private static IEnumerable<ODataExpandAssociation> ExtractExpandAssociations(MemberExpression memberExpression, ITypeCache typeCache)
+        {
+            var memberName = typeCache.GetMappedName(memberExpression.Expression.Type, memberExpression.Member.Name);
+
+            return memberExpression.Expression is MemberExpression
+                ? ExtractExpandAssociations(memberExpression.Expression, typeCache)//.Select(x => string.Join("/", x, memberName))
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs]: The commented code is removed

File: src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace Simple.OData.Client
+{
+    public static class ExpandExpression
+    {
+        public static IEnumerable<ODataExpandAssociation> ExtractExpandAssociations<T>(this Expression<Func<T, object>> expression,
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/Expressions/ExpandExpression.cs]: In my last commit I have replaced the recursion by expression visitor.

File: src/Simple.OData.Client.Core/ODataExpandAssociation.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+            var expandAssociations = new List<ODataExpandAssociation>(result.ExpandAssociations);
+            result.ExpandAssociations.Clear();
+            foreach (var association in expandAssociations)
+            {
+                result.ExpandAssociations.Add(second.ExpandAssociations.Aggregate(association, MergeExpandAssociations));
+            }
+            foreach (var association in second.ExpandAssociations)
+            {
+                if (result.ExpandAssociations.All(a => a.Name != association.Name))
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/ODataExpandAssociation.cs]: `result.ExpandAssociations.All(a => a.Name != association.Name)` and `!result.ExpandAssociations.Any(a => a.Name == association.Name)` produce the same result.  My Rider (or ReSharper) suggests replacement of check with `.Any(...)` by check with `.All(...)`

And to my knowledge are the names of properties in OData case sensitive.

File: src/Simple.OData.Client.Core/ODataExpandAssociation.cs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+            
+            return result;
+        }
+
+        public static ODataExpandAssociation From(string association)
+        {
+            if (string.IsNullOrEmpty(association))
+                return new ODataExpandAssociation(string.Empty);
+            
+            var items = association.Split('/');
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/ODataExpandAssociation.cs:26]: `association.Split('/')` can only produce empty array if the parameter `association` is empty. I check already this parameter and return an association with an empty name.

File: src/Simple.OData.Client.Core/ODataExpandAssociation.cs [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+            }
+
+            return expandAssociation;
+        }
+
+        public ODataExpandAssociation Clone()
+        {
+            var clone = new ODataExpandAssociation(Name);
+            clone.ExpandAssociations.AddRange(ExpandAssociations.Select(a => a.Clone()));
+            clone.FilterExpression = FilterExpression;
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/ODataExpandAssociation.cs:42]: ODataExpression is immutable, so I think it is safe to use reference to original filter in the clone

File: src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs (hunk 2 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs:471]: sorry, I was wrong. I will revert my changes and fix this test.

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs (hunk 3 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs]: You're right, the .ToList() call is not necessary here. I will remove this call.

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs (hunk 5 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs]: I have replaced the calls of .Skip(1).ToArray() with iteration over the segments

File: src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs (hunk 6 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs:210]: I have replaced the calls of .Skip(1).ToArray() with iteration over the segments

File: src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs (hunk 2 above)
Comment (maxbruecken) [reply src/Simple.OData.Client.UnitTests/Core/ExpansionTests.cs]: I have fixed this test, but I had to change the expected command text. According to the OData specification "A property MUST NOT appear in more than one expand item" all expand associations from each level will be merged.

File: src/Simple.OData.Client.Core/ODataExpandAssociation.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+        public static ODataExpandAssociation MergeExpandAssociations(ODataExpandAssociation expandAssociation, string path)
+        {
+            return MergeExpandAssociations(expandAssociation, From(path));
+        }
+        
+        public static ODataExpandAssociation MergeExpandAssociations(ODataExpandAssociation first, ODataExpandAssociation second)
+        {
+            var result = first.Clone();
+            if (result.Name != second.Name && result.Name != "*") return result;
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/ODataExpandAssociation.cs]: yes, the names are case sensitive

File: src/Simple.OData.Client.Core/ODataOrderByColumn.cs [hunk 17]
```diff
@@ -0,0 +1,14 @@
+namespace Simple.OData.Client
+{
+    public class ODataOrderByColumn
+    {
+        public string Name { get; }
+        public bool Descending { get; }
+
+        public ODataOrderByColumn(string name, bool descending)
+        {
+            Name = name;
```
[the remark this answers is not in the document]
Comment (maxbruecken) [reply src/Simple.OData.Client.Core/ODataOrderByColumn.cs:14]: I have added validation of the parameter 'name'

**Comment by object** [comment]
Verified the tests, merging it.
